### PR TITLE
firmware swapping ported for nucleo-f411

### PR DESCRIPTION
**Pull request by jbeyerstedt**
This should work the same way on all STM32F4 processors, if there is enough flash memory available.

This also adds a driver for sectored flash used in STM32F2, F4 and F7 microcontrollers. But the current implementation is only tested on STM32F411, but should work on all F4 nucleo boards.

### CODE REVIEW

**Comment by kYc0o** [comment]
Hi Jannik! Thanks for the PR.

I was looking at it and there are some things that we need to change, but it looks already good!

As you can see, some changes are on-going, so what I propose is maybe to split this PR into several parts, or, if you prefer, I can chery-pick the commits which are mergeable now and work on the rest later.

Besides, I looked at you PR in RIOT upstream and I think it will difficult to merge it, as well as in this design I use RAM buffers to write the flash pages, which already quite large (2K per page on smt32f103re).

So, in short, what I would propose is to take the parts which involve the configuration defines for stm32f4 and integrate them in a single PR which will include both defines, stm32f4 and stm32f1. What do you think?

I'm investigating the thing about the 512bytes of metadata so maybe I'll make this general for all platforms.

By the way, I already have a solution for OTA programming which I'll release soon, in case of you're working now on something similar.

Thanks again!

File: cpu/stm32f4/ldscripts/stm32f411re_slot1.ld [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * @brief           Memory definitions for the STM32F103RE Slot 0
+ *
+ * @author          Francisco Acosta <francisco.acosta@inria.fr>
+ *
+ * @}
+ */
+
+MEMORY
+{
+    rom (rx)    : ORIGIN = 0x08004200, LENGTH = 0x1C000
```
Comment (kYc0o) [inline cpu/stm32f4/ldscripts/stm32f411re_slot1.ld]: Why did you choose a different value for each slot?

File: examples/firmware_swapping/Makefile [hunk 2]
```diff
@@ -4,28 +4,53 @@ export BOARD ?= iotlab-m3
...
+endif
+
+metadata-generator:
+	@cd ../../dist/tools/firmware_metadata; \
+	CFLAGS="-DFW_METADATA_SPACE=$(FW_METADATA_SPACE)" make clean all
+
 bootloader:
 	@cd ../bootloader; \
-	CFLAGS+=-DFW_METADATA_SPACE=$(FW_METADATA_SPACE) make clean all; \
+	CFLAGS="-DFW_METADATA_SPACE=$(FW_METADATA_SPACE)" make clean all; \
```
Comment (kYc0o) [inline examples/firmware_swapping/Makefile]: Why did you change this? Anyways, this Makefile will disappear.

File: tests/periph_flashsector/Makefile [hunk 3]
```diff
@@ -0,0 +1,9 @@
+APPLICATION = periph_flashpage
+BOARD ?= nucleo-f411
+include ../Makefile.tests_common
+
+FEATURES_REQUIRED = periph_flashsector
+
+USEMODULE += shell
```
Comment (kYc0o) [inline tests/periph_flashsector/Makefile]: This is a no-go, if you want to include modules you need to do it at the application level. Otherwise you can also do it at the dependency check level.

**Review by kYc0o — COMMENTED** [review]
Small review, but as I said a big part of this code has changed so it won't be mergeable as is.

File: cpu/stm32f4/ldscripts/stm32f411re_slot1.ld (hunk 1 above)
Comment (jbeyerstedt) [inline cpu/stm32f4/ldscripts/stm32f411re_slot1.ld]: I have different sizes for each slot, because the sectored flash layout of the STM32F4 makes it impossible to have firmware slots with the same size.
The first slot will be smaller by the size of the bootloader space.

File: examples/firmware_swapping/Makefile (hunk 2 above)
Comment (jbeyerstedt) [inline examples/firmware_swapping/Makefile]: I changed this, because I had issues with your makefile. I used an Ubuntu 16.04 system to compile my RIOT projects.

**Comment by jbeyerstedt** [comment]
As discussed in my pull request for the flash interface of "the other half" of the STM32 micro controllers, it would be a bit tricky to integrate these two types of flash memory.

> As you can see, some changes are on-going, so what I propose is maybe to split this PR into several parts, or, if you prefer, I can chery-pick the commits which are mergeable now and work on the rest later.
So, in short, what I would propose is to take the parts which involve the configuration defines for stm32f4 and integrate them in a single PR which will include both defines, stm32f4 and stm32f1. What do you think?

You will have a better overview of you work and the longer experience developing for RIOT, so do what you think would be the best for the further development.


By the way: I will not have quite a lot of time for further development of the compatible version, because my thesis has first priority. And the prototype I’m developing for my thesis has moved to a quite different approach, which will not be (easily) mergeable.
